reconcile/test/test_deadmanssnitch.py

```diff
@@ -0,0 +1,333 @@
+                token="test_token",
+                status="healthy",
+                alert_email=["test_email"],
+                notes="test_notes",
+                check_in_url="testURL2",
+                interval="15_minute",
+                href="test_href",
+                alert_type="test_type",
+                tags=["appsre"],
+                vault_data="testURL",
+            ),
+            "test_cluster_5": Snitch(
+                name="test_cluster_5",
+                token="test_token",
+                status="healthy",
+                alert_email=["test_email"],
+                notes="test_notes",
+                check_in_url="testURL",
+                interval="15_minute",
+                href="test_href",
+                alert_type="test_type",
+                tags=["appsre"],
+                vault_data="testURL2",
+            ),
+        },
+        desired_state=[
+            ClusterV1(
+                name="test_cluster_1",
+                serverUrl="testurl",
+                consoleUrl="test_c_url",
+                alertmanagerUrl="test_alert_manager",
+                managedClusterRoles=True,
+                prometheusUrl="test_prom_url",
+                enableDeadMansSnitch=True,
+            ),
+            ClusterV1(
+                name="test_cluster_3",
+                serverUrl="testurl",
+                consoleUrl="test_c_url",
+                alertmanagerUrl="test_alert_manager",
+                managedClusterRoles=True,
+                prometheusUrl="test_prom_url",
+                enableDeadMansSnitch=False,
+            ),
+            ClusterV1(
+                name="test_cluster_new",
+                serverUrl="testurl",
+                consoleUrl="test_c_url",
+                alertmanagerUrl="test_alert_manager",
+                prometheusUrl="test_prom_url",
+                managedClusterRoles=True,
+                enableDeadMansSnitch=True,
+            ),
+            ClusterV1(
+                name="test_cluster_5",
+                serverUrl="testurl",
+                consoleUrl="test_c_url",
+                alertmanagerUrl="test_alert_manager",
+                prometheusUrl="test_prom_url",
+                managedClusterRoles=True,
+                enableDeadMansSnitch=True,
+            ),
+        ],
+    )
+    expected_output = [
+        DiffData(
+            cluster_name="test_cluster_new",
+            action=Action.create_snitch,
+        ),
+        DiffData(
+            cluster_name="test_cluster_3",
+            data="test_token",
+            action=Action.delete_snitch,
+        ),
+        DiffData(
+            cluster_name="test_cluster_5",
+            data="testURL",
+            action=Action.update_vault,
+        ),
+    ]
+
+    expected_output_map = [data.__dict__ for data in expected_output]
+    diff_data_map = [data.__dict__ for data in diff_data]
+    assert diff_data_map == expected_output_map
```

Comment:
why require `__dict__`? If `DiffData` is a `BaseModel`, then eq compare should work for value check

Reply:
@hemslo `DiffData` is not a BaseModel so had to use `__dict__` for testcase

Reply:
Can we make it a BaseModel to simplify this, since it's a dataclass, either use `@dataclass` or BaseModel to make it explicit.

Reply:
Now that it becomes a `BaseModel`, it reveals a type check error, `data` can be optional, but when action is delete, `data` can't be None. Can describe this kind of typing strictly like this:

```python
class DiffData(BaseModel, ABC):
    cluster_name: str
    action: Action

class CreateSnitchDiffData(DiffData):
    action = Action.create_snitch

class DeleteSnitchDiffData(DiffData):
    action = Action.delete_snitch
    data: str

class UpdateVaultDiffData(DiffData):
    action = Action.update_vault
    data: str
```

You can rename `data` to better suit for the meaning of that class. Then you can simply use the subclass directly and pattern match on it:

```python
        create_snitches = [
            CreateSnitchDiffData(cluster_name=cluster)
            # add cluster to create set only in case they do not exist in current state
            for cluster in desired_cluster_enabled_true - current_state.keys()
        ]
```

```python
    def apply_diff(self, diff: DiffData) -> None:
        match diff:
            case CreateSnitchDiffData(cluster_name=cluster_name):
                self.create_snitch(cluster_name)
```